proteny
=======

A tool to analyze synteny at the protein level.
We develop an algorithm to detect statistically significant clusters of exons between two proteomes.
The tool provides algorithms to quickly detect and visualize results in order to support conclusions from genomic data.

The method discovers clusters of hits from a bi-directional BLASTp of translated exon sequences in two organisms.
A dendrogram for hits is built based on genomic distances between hits, and cut based on significance of a cluster score based on a permutation test at each node in the tree.
The result is a set of large clusters describing high exonic conservation.

![An example of the figures generated by proteny](/readme/example_output.gif)

Algorithm
=========

We use BLASTp to produce a set of hits, which are used to build

![We use BLASTp to produce a set of hits, which are used to build](/readme/clustering_dendrogram_a.gif)

a dendrogram which is traversed to find

![a dendrogram which is traversed to find](/readme/clustering_dendrogram_b.gif)

significant clusters

![significant clusters.](/readme/clustering_dendrogram_c.gif)

Which are then visualized in a useful way

![Visualizations with Circos](/readme/visualization.gif)


Dependencies
=============
 * CIRCOS: http://circos.ca/
 * IBIDAS: https://github.com/mhulsman/ibidas
 * BLAST+: ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/

Usage
=========

Make sure that each dependency (and their dependencies) is installed.
modify the "$CIRCOS" variable in "run" to reflect the location of your circos installation.

"example.py" contains an example which details the way the tool can be used.
"run" is a script which runs circos finding each configuration file, to produce the PNG and SVG files.

Although the ipython 'ipcluster' is started automatically by Proteny when it is needed, it does not always behave nicely.
I therefore recommend starting the ipcluster before starting proteny.

```shell
$> ipcluster -n 8 # Start the ipython cluster manager
$> ipython
In [1]: execfile("example.py");
```

Running the example
=====================

You need to download some files from the JGI DOE in order to run this example.
The files are accessible for free, but require an account:

 * http://genome.jgi-psf.org/Schco2/download/Schco2_AssemblyScaffolds.fasta.gz
 * http://genome.jgi-psf.org/Schco2/download/Schco2_GeneCatalog_genes_20110923.gff.gz

 * http://genome.jgi-psf.org/Agabi_varbisH97_2/download/Abisporus_varbisporusH97.v2.maskedAssembly.gz
 * http://genome.jgi-psf.org/Agabi_varbisH97_2/download/Abisporus_var_bisporus.mitochondrion.scaffolds.fasta.gz
 * http://genome.jgi-psf.org/Agabi_varbisH97_2/download/Abisporus_varbisporusH97.v2.FilteredModels3.gff.gz

Then, you can run the example as usual:

```shell
$> ipcluster -n 8 # Start the ipython cluster manager
$> ipython
In [1]: execfile("example.py");
```

